src/sqlfluff/core/templaters/python.py

```diff
@@ -965,20 +976,59 @@ def _split_uniques_coalesce_rest(
                         # Anything up to start_idx (exclusive) is included.
                         include_start = raw_idx > elem_sub_buffer[0][2]
 
-                        # The ending point of this slice, is already decided.
-                        end_point = elem_sub_buffer[-1].end_source_idx()
-
                         # If start_idx is None, we're in luck. We don't need to include the beginning.
                         if include_start:
                             start_point = elem_sub_buffer[0].source_idx
                         # Otherwise we know it's looped round, we need to include the whole slice.
                         else:
                             start_point = elem_sub_buffer[cur_idx].source_idx
 
+                        # The ending point of this slice, is already decided.
+                        end_point = elem_sub_buffer[-1].end_source_idx()
+
+                        # However if can match beginning as a literal, then let's split
+                        # to improve some (but not all) for loop use cases
+                        if elem_sub_buffer[0].raw == int_file_slice.slice_buffer[0].raw:
+                            end_point = elem_sub_buffer[0].end_source_idx()
+
+                            # It might be the whole beginning is a literal (though
+                            # think this should be covered before we get here)
+                            # If not then yield the first literal, then set up the
+                            # slice points for the second, non-literal part.
+                            # Otherwise we can fall out to the final yield as
+                            # slice points are already set above.
+                            if end_point < elem_sub_buffer[-1].end_source_idx():
+                                tricky_initial_literal = TemplatedFileSlice(
```

Comment:
Would it be reasonable here to include a brief example of this tricky case? When working in unfamiliar code, it really helps my understanding to see examples alongside the code for handling them.